src/Bicep.Core/Syntax/SyntaxRewriteVisitor.cs

```diff
@@ -0,0 +1,534 @@
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            // TODO calculate value correctly here
+            return new BooleanLiteralSyntax(literal, syntax.Value);
+        }
+        void ISyntaxVisitor.VisitBooleanLiteralSyntax(BooleanLiteralSyntax syntax) => ReplaceCurrent(syntax, ReplaceBooleanLiteralSyntax);
+
+        protected virtual StringSyntax ReplaceStringSyntax(StringSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.StringTokens, out var stringTokens);
+            hasChanges |= Rewrite(syntax.Expressions, out var expressions);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            // TODO calculate segmentValues correctly here
+            return new StringSyntax(stringTokens, expressions, syntax.SegmentValues);
+        }
+        void ISyntaxVisitor.VisitStringSyntax(StringSyntax syntax) => ReplaceCurrent(syntax, ReplaceStringSyntax);
+
+        protected virtual ProgramSyntax ReplaceProgramSyntax(ProgramSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.Children, out var children);
+            hasChanges |= Rewrite(syntax.EndOfFile, out var endOfFile);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            // TODO what to do with lexerDiagnostics?
+            return new ProgramSyntax(children, endOfFile, syntax.LexerDiagnostics);
+        }
+        void ISyntaxVisitor.VisitProgramSyntax(ProgramSyntax syntax) => ReplaceCurrent(syntax, ReplaceProgramSyntax);
+
+        protected virtual NumericLiteralSyntax ReplaceNumericLiteralSyntax(NumericLiteralSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.Literal, out var literal);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            // TODO calculate value correctly here
+            return new NumericLiteralSyntax(literal, syntax.Value);
+        }
+        void ISyntaxVisitor.VisitNumericLiteralSyntax(NumericLiteralSyntax syntax) => ReplaceCurrent(syntax, ReplaceNumericLiteralSyntax);
+
+        protected virtual NullLiteralSyntax ReplaceNullLiteralSyntax(NullLiteralSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.NullKeyword, out var nullKeyword);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            return new NullLiteralSyntax(nullKeyword);
+        }
+        void ISyntaxVisitor.VisitNullLiteralSyntax(NullLiteralSyntax syntax) => ReplaceCurrent(syntax, ReplaceNullLiteralSyntax);
+
+        protected virtual SkippedTriviaSyntax ReplaceSkippedTriviaSyntax(SkippedTriviaSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.Elements, out var elements);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            // TODO what to do with span & diagnostics?
```

Comment:
>span [](start = 36, length = 4)

Not sure if this helps. The span is only really needed to figure out the position for zero-element cases. Otherwise it can be calculated from elements.

Reply:
On the diagnostics front, I think we should clear them because the syntax node got rewritten and they may no longer apply. This leaves situations where we could lose some errors, but I think it should be on the particular rewriter implementation to not rewrite code into something invalid.

---
In reply to: [527481476](https://github.com/Azure/bicep/pull/953#discussion_r527481476) [](ancestors = 527481476)

Reply:
Yeah I think I agree. I think if rewriting syntax, it's logical that you have to re-run the compiler to get access to meaningful errors.